CRM/ManageLetterheads/Hook/BuildForm/AddLetterheadSelect.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+/**
+ * Adds the letterhead select element to the Email and PDF Letter forms.
+ *
+ * The list of letterheads that should be visible for each form is also appended
+ * as a configuration variable.
+ */
+class CRM_ManageLetterheads_Hook_BuildForm_AddLetterheadSelect {
+
+  /**
+   * @var bool
+   */
+  private $isEmailForm = FALSE;
+
+  /**
+   * @var bool
+   */
+  private $isPdfLetterForm = FALSE;
+
+  /**
+   * Build Form hook handler.
+   *
+   * @param string $formName
+   *   The name for the current form.
+   * @param CRM_Core_Form $form
+   *   The curernt form instance.
+   */
+  public function run ($formName, $form) {
+    $this->isPdfLetterForm = get_class($form) === CRM_Contact_Form_Task_PDF::class;
+    $this->isEmailForm = get_class($form) === CRM_Contact_Form_Task_Email::class;
+
+    if (!$this->shouldRun()) {
+      return;
+    }
+
+    $availableForName = $this->isEmailForm ? 'emails' : 'pdf_letters';
+    $availableForValue = $this->getAvailableForValueFromName($availableForName);
+    $letterheads = $this->getLetterheadsFromAvailableFor($availableForValue);
+
+    $this->addListOfLetterheadOptionsToConfig($letterheads);
+  }
+
+  /**
+   * Returns the "Available For" option value for the given name.
+   *
+   * @param string $optionName
+   *   The name of the "Available For" option to return.
+   * @return array
+   */
+  private function getAvailableForValueFromName($optionName) {
+    $optionValue = civicrm_api3('OptionValue', 'getsingle', [
+      'is_sequential' => '1',
+      'option_group_id' => 'manageletterheads_available_for',
+      'name' => $optionName,
+    ]);
+
+    return $optionValue['value'];
+  }
+
+  /**
+   * Returns all the letterheads that belong to the given "Available For" value.
+   *
+   * Only active letterheads are returned, sorted by their weight.
+   *
+   * @param string $availableForValue
+   *   The "Available For" value to filter letterheads by.
+   */
+  private function getLetterheadsFromAvailableFor($availableForValue) {
```

Comment:
I think we can simplify this function, Instead of looping through all the letterheads and then filtering out which one belongs to Emails and which one belongs to PDF letters.  
We can also do away with `getAvailableForValueFromName` function.

```php

//Return a list of all letterhead ID's for Emails
$result = civicrm_api3('LetterheadAvailability', 'get', [
  'return' => ['letterhead_id'],
  'available_for' => 'emails', // 'available_for' => 'pdf_letters'
]);


$letterHeadIds = array_column($result['values'], 'letterhead_id'));

$result = civicrm_api3('Letterhead', 'get', [
  'id' => ['IN' => $letterHeadIds],
  'is_active' => '1',
  'options' => ['sort' => 'weight ASC'],
]);


return $result['values']

```